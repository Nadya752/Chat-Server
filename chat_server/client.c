#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <dirent.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <dirent.h>
#include "client.h"

// This file is for testing purposes only.

// Creates messages of client when they wants to SAYCONT in testing.
void saycont(char* id_name, char* domain_name, char* newmsg, int num){

    // Testing client will say "Hello I'm <id>", 
    // "Nice to meet you all", "I'm from <domain>".
    unsigned char type = SAYCONT;
    newmsg[0] = type;
    newmsg[1] =  0;

    if (num == 0){
        strcpy((char*) newmsg+START_PKT, "Hello I'm ");
        strcat((char*) newmsg+MSG_LEN1, id_name);
    }else if (num == 1){
        strcpy((char*) newmsg+START_PKT, "Nice to meet you all");
    }else {
        strcpy((char*) newmsg+START_PKT, "I'm from ");
        strcat((char*) newmsg+MSG_LEN2, domain_name);
        newmsg[MAX_BUF-1] = (char ) TERM;
    }

}

// Creates messages of client when they wants to SAY in testing.
void say(char* id_name, char* newmsg){

    // Testing client will say "Hello I'm <id>".
    unsigned char type = SAY;
    newmsg[0] = type;
    newmsg[1] =  0;
    strcpy(newmsg+START_PKT, "Hello I'm ");
    strcat(newmsg+MSG_LEN1, id_name);
}

// Creates messages of client when they wants to CONNECT.
void connect(char* id_name, char* domain_name, char* newmsg){
   unsigned char type = CONNECT;

    newmsg[0] = type;
    newmsg[1] =  0;
    short nidx = START_PKT;
    short ididx = 0;
    short didx = 0;
    while (id_name[ididx] != '\0' && ididx<= MAX_ID){
        newmsg[nidx] = id_name[ididx];
        nidx++;
        ididx++;

    }

    nidx = END_ID;
    while (domain_name[didx] != '\0' && didx <= (MAX_BUF - MAX_ID - START_PKT)){
        newmsg[nidx] = domain_name[didx];
        nidx++;
        didx++;

    }

}

// Checks if gevent pipe exists.
int check_gevent(){
    DIR* dir;
    struct dirent* folder;
    char filename[MAX_PNAME];
    char pipe_name[] = "gevent";
    dir = opendir(".");

    if (dir){
        while((folder= readdir(dir)) != NULL){
            memset(filename, 0,MAX_PNAME* sizeof(filename[0]));
            strcpy(filename, folder->d_name);
            if (strcmp(filename, pipe_name) == 0){
                printf("Pipe gevent is open\n");
                return TRUE;
            }
        }

    }
    printf("Pipe gevent is missing\n");
    return FALSE;
}

// Checks if a given domain exists.
int check_domain(char* dom){
    DIR* dir;
    dir = opendir(dom);

    if (dir){
        closedir(dir);
        printf("Domain %s exists\n", dom);
        return TRUE;
    }
    printf("Domain %s does not exists\n", dom);
    return FALSE;
}

// Checks if _RD and _WR pipes of given id in domain exists.
int check_rd_wr_pipe(char* dom, char* pipe_rd, char* pipe_wr){
    int has_pread = FALSE;
    int has_pwrite = FALSE;
    DIR* dir;
    struct dirent* folder;
    char filename[MAX_PNAME];
    dir = opendir(dom);

    // Open domain folder and search pipes.
    if (dir){
        while((folder= readdir(dir)) != NULL){
            memset(filename, 0,MAX_PNAME* sizeof(filename[0]));
            strcpy(filename, folder->d_name);
            if (strcmp(filename, pipe_rd) == 0){
                printf("Pipe %s is open\n", pipe_rd);
                has_pread = TRUE;
            }else if (strcmp(filename, pipe_wr)== 0){
                printf("Pipe %s is open\n", pipe_wr);
                has_pwrite = TRUE;
            }
        }

    }else{
        printf("failed opening domain\n");
    }
    
    return (has_pread && has_pwrite);

}
