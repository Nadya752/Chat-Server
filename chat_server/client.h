#ifndef CLIENT_H
#define CLIENT_H

#define MAX_ID (256)
#define TERM (255)
#define MAX_BUF (2048)
#define END_ID (258)
#define START_PKT (2)
#define MAX_PNAME (262)
#define MSG_LEN1 (10)
#define MSG_LEN2 (11)

typedef enum{
    CONNECT,
    SAY,
    SAYCONT,
    RECEIVE,
    RECVCONT,
    PING,
    PONG,
    DISCONNECT

}protocol;

typedef enum{
    FALSE,
    TRUE
    
} boolean;

int check_gevent();
int check_domain(char* dom);
int check_rd_wr_pipe(char* dom, char* pipe_rd, char* pipe_wr);
void connect(char* id_name, char* domain_name, char* newmsg);
void say(char* id_name, char* newmsg);
void saycont(char* id_name, char* domain_name, char* newmsg, int num);

#endif
