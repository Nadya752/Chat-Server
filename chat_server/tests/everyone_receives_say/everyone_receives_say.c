#include "../../client.h"
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <dirent.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include <dirent.h>

// Checks if everyone in the domain receives SAY message.
int main(){

    // Client named Ron connects to Gryffindor.
    char conmsg1[MAX_BUF] = {0};
    connect("Ron", "Gryffindor", conmsg1);
    int fdcon1 = open("gevent", O_WRONLY);
    if (fdcon1 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon1, conmsg1, MAX_BUF);
    close(fdcon1);

    // Client named Harry connects to Gryffindor.
    char conmsg2[MAX_BUF] = {0};
    connect("Harry", "Gryffindor", conmsg2);
    int fdcon2 = open("gevent", O_WRONLY);
    if (fdcon2 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon2, conmsg2, MAX_BUF);

    // Client named Hermione connects to Gryffindor.
    char conmsg3[MAX_BUF] = {0};
    connect("Hermione", "Gryffindor", conmsg3);
    int fdcon3 = open("gevent", O_WRONLY);
    if (fdcon3 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon3, conmsg3, MAX_BUF);

    // Client named Fred connects to Gryffindor.
    char conmsg4[MAX_BUF] = {0};
    connect("Fred", "Gryffindor", conmsg4);
    int fdcon4 = open("gevent", O_WRONLY);
    if (fdcon4 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon4, conmsg4, MAX_BUF);

    // Client named George connects to Gryffindor.
    char conmsg5[MAX_BUF] = {0};
    connect("George", "Gryffindor", conmsg5);
    int fdcon5 = open("gevent", O_WRONLY);
    if (fdcon5 == -1){
        printf("Failed opening gevent.\n");
    }
    write(fdcon5, conmsg5, MAX_BUF);

    sleep(2);
    // Ron SAYs his message.
    char saymsg[MAX_BUF] = {0};
    say("Ron", saymsg);
    for ( int idx = 0; idx<5; idx++){
        ssize_t pwrite = open("./Gryffindor/Ron_WR", O_WRONLY);
        int wrres = write(pwrite,saymsg, MAX_BUF);
        if(wrres == -1){
            fprintf(stderr, "%s", "Can't write pipe ");
        }
        if (pwrite == -1){
            perror("open()");
            fprintf(stderr, "%s", "Can't open write pipe\n");
        }
        close(pwrite);

    }


    sleep(2);
    // Checks if Hermione receives the message.
    char newmsg[MAX_BUF] = {0};
    ssize_t pread = open("./Gryffindor/Hermione_RD", O_RDONLY);
    read(pread,newmsg, MAX_BUF);
    for(int idx = START_PKT; idx < MAX_BUF; idx++){
        if (newmsg[idx] != 0){
            printf("%c", newmsg[idx]);
        }

    }
    close(pread);
    printf("\n");

    // Checks if Harry receives the message.
    char newmsg2[MAX_BUF] = {0};
    ssize_t pread2 = open("./Gryffindor/Harry_RD", O_RDONLY);
    read(pread2,newmsg2, MAX_BUF);
    for(int idx = START_PKT; idx < MAX_BUF; idx++){
        if (newmsg2[idx] != 0){
            printf("%c", newmsg2[idx]);
        }

    }
    close(pread2);
    printf("\n");

    // Checks if Fred receives the message.
    char newmsg3[MAX_BUF] = {0};
    ssize_t pread3 = open("./Gryffindor/Fred_RD", O_RDONLY);
    read(pread3,newmsg3, MAX_BUF);
    for(int idx = START_PKT; idx < MAX_BUF; idx++){
        if (newmsg3[idx] != 0){
            printf("%c", newmsg3[idx]);
        }

    }
    close(pread3);
    printf("\n");

    // Checks if George receives the message.
    char newmsg4[MAX_BUF] = {0};
    ssize_t pread4 = open("./Gryffindor/George_RD", O_RDONLY);
    read(pread4,newmsg4, MAX_BUF);
    for(int idx = START_PKT; idx < MAX_BUF; idx++){
        if (newmsg4[idx] != 0){
            printf("%c", newmsg4[idx]);
        }

    }
    close(pread4);
    printf("\n");

    return 0;

}