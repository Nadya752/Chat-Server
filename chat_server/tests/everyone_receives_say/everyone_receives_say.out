RonHello I'm Ron
RonHello I'm Ron
RonHello I'm Ron
RonHello I'm Ron
